src/app/codegen-data-model/CodegenDataModel_Write.cpp

```diff
@@ -0,0 +1,401 @@
+
+        VerifyOrReturnError(!(*attributeMetadata)->MustUseTimedWrite() ||
+                                request.writeFlags.Has(InteractionModel::WriteFlags::kTimed),
+                            CHIP_IM_GLOBAL_STATUS(NeedsTimedInteraction));
+    }
+
+    // Extra check: internal requests can bypass the read only check, however global attributes
+    // have no underlying storage, so write still cannot be done
+    VerifyOrReturnError(attributeMetadata != nullptr, CHIP_IM_GLOBAL_STATUS(UnsupportedWrite));
+
+    if (request.path.mDataVersion.HasValue())
+    {
+        std::optional<InteractionModel::ClusterInfo> clusterInfo = GetClusterInfo(request.path);
+        if (!clusterInfo.has_value())
+        {
+            ChipLogError(DataManagement, "Unable to get cluster info for Endpoint 0x%x, Cluster " ChipLogFormatMEI,
+                         request.path.mEndpointId, ChipLogValueMEI(request.path.mClusterId));
+            return CHIP_IM_GLOBAL_STATUS(DataVersionMismatch);
+        }
+
+        if (request.path.mDataVersion.Value() != clusterInfo->dataVersion)
+        {
+            ChipLogError(DataManagement, "Write Version mismatch for Endpoint 0x%x, Cluster " ChipLogFormatMEI,
+                         request.path.mEndpointId, ChipLogValueMEI(request.path.mClusterId));
+            return CHIP_IM_GLOBAL_STATUS(DataVersionMismatch);
+        }
+    }
+
+    AttributeAccessInterface * aai       = GetAttributeAccessOverride(request.path.mEndpointId, request.path.mClusterId);
+    std::optional<CHIP_ERROR> aai_result = TryWriteViaAccessInterface(request.path, aai, decoder);
+    if (aai_result.has_value())
+    {
+        if (*aai_result == CHIP_NO_ERROR)
+        {
+            // TODO: change callbacks should likely be routed through the context `MarkDirty` only
+            //       however for now this is called directly because ember code does this call
+            //       inside emberAfWriteAttribute.
+            MatterReportingAttributeChangeCallback(request.path);
+            CurrentContext().dataModelChangeListener->MarkDirty(request.path);
+        }
+        return *aai_result;
+    }
+
+    MutableByteSpan dataBuffer = gEmberAttributeIOBufferSpan;
+    ReturnErrorOnFailure(DecodeValueIntoEmberBuffer(decoder, *attributeMetadata, dataBuffer));
+
+    Protocols::InteractionModel::Status status;
+
+    if (request.operationFlags.Has(InteractionModel::OperationFlags::kInternal))
+    {
+        // Internal requests use the non-External interface that has less enforcement
+        // than the external version (e.g. does not check/enforce writable settings, does not
+        // validate attribute types) - see attribute-table.h documentation for details.
+        status = emberAfWriteAttribute(request.path.mEndpointId, request.path.mClusterId, request.path.mAttributeId,
+                                       gEmberAttributeIOBufferSpan.data(), (*attributeMetadata)->attributeType);
```

Comment:
Also, shouldn't we be using `dataBuffer.data()` in some of these places?  The use of `gEmberAttributeIOBufferSpan` when `dataBuffer` is what we decoded the value into is a little weird, even though they do happen to point to the same data....

Reply:
Yes, forgot to properly update.